### DISCUSSION TITLE
Comparison Between getFullList and getList Methods in PocketBase SDK - Unexpected Results

### DISCUSSION THREAD

**Discussion by vladimirzb** [pocketbase/pocketbase#3072] [Q&A] [answered]
Hello,

I'm working with the PocketBase SDK in TypeScript and have encountered a situation where I might be doing something wrong. I'm trying to understand the difference between the `getFullList` and `getList` methods.

Here's the code snippet for `getFullList`:

```typescript
async function obtenerIngresos(): Promise<IngresoRecord[]> {
    try {
        const records = await pb.collection('ingresos').getFullList({
            sort: '-created',
        });
        console.log("With getFullList:", records); // Log the full records
        return records as unknown as IngresoRecord[]; // Type assertion here
    } catch (error) {
        console.error('Failed to fetch data:', error);
        throw error;
    }
}
```

Result with `getFullList`:

```plaintext
[
  Record {
    cantidad: 11111,
    collectionId: 'el5tmq7mxxwvyyi',
    collectionName: 'ingresos',
    created: '2023-07-23 20:35:02.722Z',
    fecha: '2023-07-05 12:00:00.000Z',
    fuente: 'aaa',
    id: '72ay6589m0fp9h1',
    updated: '2023-07-23 20:35:02.722Z',
    expand: {}
  }
]
```

Here's the code snippet for `getList`:

```typescript
async function obtenerIngresos(): Promise<IngresoRecord[]> {
    try {
        const resultList = await pb.collection('ingresos').getList(1, 50, {
            filter: 'created >= "2022-01-01 00:00:00"',
        });
        console.log("With getList:", resultList.items); // Log the full result list
        return resultList.items;
    } catch (error) {
        console.error('Failed to fetch data:', error);
        throw error;
    }
}
```

Result with `getList`:

```plaintext
[
  Record { /* Record 1 details */ },
  Record { /* Record 2 details */ },
  Record { /* Record 3 details */ }
]
```

As per the documentation, I expected the `getFullList` method to fetch all records at once, similar to `getList`. However, `getFullList` is only returning one record, while `getList` returns all the records as expected.

I'm not sure if I'm misunderstanding the usage of these methods or if there's something specific I need to do differently. Could you please help me understand what might be going wrong?

Thank you for your assistance!

**Comment by ganigeorgiev** [comment]
What versions of the JS SDK and PocketBase executable do you use?

Also keep in mind that the result of the 2 calls may differ depending on whether you are authenticated and what records the request client have access to based on the List API rule of the collection.

_As a side-note, you can pass a generic type to the functions instead of manually performing the type assertion, aka. `getFullList<IngresoRecord>(...)`._

**Reply by vladimirzb** [reply]
Thank you for your prompt response. I'm providing the information you requested to understand the unexpected results between `getFullList` and `getList`:

SDK Version: 0.15.3  
PocketBase version: v0.17.2 Release  
Image of access control: 
![image](https://github.com/pocketbase/pocketbase/assets/55926702/5beb15d6-67d8-41af-9dde-b0d65d4f2bd6)


I appreciate the side-note about passing a generic type to the functions. It indeed made my code cleaner, and I'm grateful for the tip.

If there's anything else you need to diagnose the issue or if you have further insights, please let me know. I'm eager to understand what might be causing this discrepancy in my code or database setup.

Thanks again for your assistance!

**Reply by ganigeorgiev** [reply, marked as answer]
@vladimirzb I'm not able to reproduce the issue locally.

If you use Next.js or other similar framework, make sure that it is not caused by a cache configuration (see https://github.com/pocketbase/pocketbase/discussions/2695).

`getFullList()` internally calls `getList()` until there are no more items to fetch, so there shouldn't be any difference in the returned results (_especially since you have an empty list rule and all clients would have access to the same records_).

Also, since you are using PocketBase v0.17+, I'd advice updating your [JS SDK to v0.16.0](https://github.com/pocketbase/js-sdk/releases) so that you can make use of the slight performance boost that comes with the default value of `skipTotal`.

If the above is not caused by a cache, please create a **minimal** reproducible repo and I'll try to investigate it further.

**Reply by vladimirzb** [reply]
@ganigeorgiev Thank you for your assistance and patience as I worked through this issue. Your recommendation to check the cache configuration in Next.js led me to the solution. I discovered that the error was due to not using `export const fetchCache = 'default-no-store';` in the page of the component. By making this adjustment, the caching issue has been resolved, and the data is now updating as expected.

This solution is applying to all the requests of the page, and I wanted to know if I could choose the cache option per request individually, similar to how it's done with the native fetch API like this: `fetch('https://...', { cache: 'no-store' })`. How can I achieve this per-request caching control with the PocketBase SDK?

**Reply by ganigeorgiev** [reply]
There is no such option at the moment. An alternative approach could be to attach to the [`pb.beforeSend` hook func](https://github.com/pocketbase/js-sdk#send-hooks) and modify the options argument, but this hook is global and will be triggered for all requests.

_In addition to `fetchCache`, you can explore other cache strategies in the [Next.js data cache docs](https://nextjs.org/docs/app/building-your-application/caching#data-cache) and more specifically [Opting out of data caching](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching)._

**Reply by vladimirzb** [reply]
@ganigeorgiev Thank you once again for your help in resolving the caching issue. For the time being, I'll be utilizing the "Segment Config Options" that apply to all fetches on the page. If necessary, I may also consider using the native fetch API to have more granular control over caching.

I've been contemplating the idea of making a "feature request" for the ability to specify cache settings per fetch with the PocketBase SDK. I believe this could be a beneficial addition for many developers who require more nuanced control over caching behavior. However, I understand that this might be outside the typical scope of the SDK.

If you think this feature aligns with the goals and direction of the PocketBase SDK, I'll proceed with submitting the request. If not, I completely understand and appreciate the existing functionality and flexibility the SDK provides.

Thank you once again for your support and guidance!

**Reply by ganigeorgiev** [reply]
> I've been contemplating the idea of making a "feature request" for the ability to specify cache settings per fetch with the PocketBase SDK.

The JS SDK is in the process of being refactored and custom per request configurations (including specifying custom fetch options or even completely different `fetch` implementation) will be available with the next JS SDK release.
